src/sentry/workflow_engine/processors/workflow.py

```diff
@@ -34,6 +36,71 @@ class WorkflowDataConditionGroupType(StrEnum):
     WORKFLOW_TRIGGER = "workflow_trigger"
 
 
+def get_actions_to_workflows(actions_to_fire: list[Action]) -> dict[int, Workflow]:
+    action_ids = {action.id for action in actions_to_fire}
+    workflow_actions = (
+        WorkflowDataConditionGroup.objects.select_related("workflow", "condition_group")
+        .filter(condition_group__dataconditiongroupaction__action_id__in=action_ids)
+        .values_list("condition_group__dataconditiongroupaction__action_id", "workflow_id")
+    )
+
+    action_to_workflows: dict[int, int] = {
+        action_id: workflow_id for action_id, workflow_id in workflow_actions
+    }
+    workflow_ids = set(action_to_workflows.values())
+
+    workflows = Workflow.objects.filter(id__in=workflow_ids)
+    workflow_ids_to_workflows = {workflow.id: workflow for workflow in workflows}
+
+    # TODO: account for actions being attached to multiple workflows?
+    # should you even send multiple notifications then? :think:
+    actions_to_workflows: dict[int, Workflow] = {
+        action_id: workflow_ids_to_workflows[workflow_id]
+        for action_id, workflow_id in action_to_workflows.items()
+    }
+    return actions_to_workflows
+
+
+def create_workflow_fire_histories(
+    actions_to_workflows: dict[int, Workflow], event_data: WorkflowEventData
+) -> dict[int, WorkflowFireHistory]:
+    workflows = set(actions_to_workflows.values())
+
+    workflow_fire_histories = [
+        WorkflowFireHistory(
+            workflow=workflow,
+            group=event_data.event.group,
+            event_id=event_data.event.event_id,
+        )
+        for workflow in workflows
+    ]
+    fire_histories = WorkflowFireHistory.objects.bulk_create(workflow_fire_histories)
+
+    workflow_id_to_fire_history: dict[int, WorkflowFireHistory] = {
+        workflow_fire_history.workflow_id: workflow_fire_history
+        for workflow_fire_history in fire_histories
+    }
+    return workflow_id_to_fire_history
+
+
+def fire_actions(actions: list[Action], detector: Detector, event_data: WorkflowEventData) -> None:
```

Comment:
not sure the correct home for this 🤔 could live in `sentry/workflow_engine/processors/action.py` too?

Reply:
yeah, i think this belongs in that file rather than here. the focus is on triggering the actions.